pkg/webhook/pod/pd_deleter.go

```diff
@@ -0,0 +1,175 @@
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+	ordinal, err := operatorUtils.GetOrdinalFromPodName(name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isMember, err := IsPodInPdMembers(tc, pod, pdClient)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isInOrdinal, err := operatorUtils.IsPodOrdinalNotExceedReplicas(pod, *ownerStatefulSet.Spec.Replicas)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isUpgrading := IsStatefulSetUpgrading(ownerStatefulSet)
+
+	IsDeferDeleting := IsPodWithPDDeferDeletingAnnotations(pod)
+
+	klog.Infof("receive delete pd pod[%s/%s] of tc[%s/%s],isMember=%v,isInOrdinal=%v,isUpgrading=%v,IsDeferDeleting=%v", namespace, name, namespace, tcName, isMember, isInOrdinal, isUpgrading, IsDeferDeleting)
+
+	// NotMember represents this pod is deleted from pd cluster or haven't register to pd cluster yet.
+	// We should ensure this pd pod wouldn't be pd member any more.
+	if !isMember {
+		return pc.admitDeleteNonPDMemberPod(IsDeferDeleting, isInOrdinal, pod, ownerStatefulSet, tc, pdClient, ordinal)
+	}
+
+	// NotInOrdinal represents this is an scale-in operation, we need to delete this member in pd cluster
+	if !isInOrdinal {
+		return pc.admitDeleteExceedReplicasPDPod(pod, tcName, pdClient)
+	}
+
+	// If there is an pd pod deleting operation during upgrading, we should
+	// check the pd pod which had been upgraded before were all health
+	if isUpgrading {
+		klog.Infof("receive delete pd pod[%s/%s] of tc[%s/%s] is upgrading, make sure former pd upgraded status was health", namespace, name, namespace, tcName)
+		err = checkFormerPDPodStatus(pc.kubeCli, pdClient, tc, namespace, ordinal, *ownerStatefulSet.Spec.Replicas)
+		if err != nil {
+			return util.ARFail(err)
+		}
+	}
+
+	leader, err := pdClient.GetPDLeader()
+	if err != nil {
+		klog.Errorf("tc[%s/%s] fail to get pd leader %v,refuse to delete pod[%s/%s]", namespace, tc.Name, err, namespace, name)
+		return util.ARFail(err)
+	}
+
+	klog.Infof("tc[%s/%s]'s pd leader is pod[%s/%s] during deleting pod[%s/%s]", namespace, tc.Name, namespace, leader.Name, namespace, name)
+	if leader.Name == name {
+		return pc.admitDeletePDLeader(pod, tc, pdClient, ordinal)
+	}
+
+	klog.Infof("pod[%s/%s] is not pd-leader,admit to delete", namespace, name)
+	return util.ARSuccess()
+}
+
+// this pod is not a member of pd cluster currently, it could be deleted from pd cluster already or haven't registered in pd cluster
+// we need to check whether this pd pod has been ensured wouldn't be a member in pd cluster
+func (pc *PodAdmissionControl) admitDeleteNonPDMemberPod(IsDeferDeleting, isInOrdinal bool, pod *corev1.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient, ordinal int32) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+
+	// check whether this pod has been ensured wouldn't be a member in pd cluster
+	if IsDeferDeleting {
```

Comment:
If we scale out and then scale in soon, the new pd pod will never be deleted. 
In addition, if a new Pod is Pending and it will never be deleted.

Reply:
The new pod would be add deferdeleting annotation and deleted memeber again. It would be deleted in next round.